### ISSUE TITLE
Unable to open zabbix_sender on discovery

### ISSUE THREAD

**Issue by chrisgassen** [takala-jp/zabbix-proxmox#26]
Hey,
we noticed this problem after our upgrade to Proxmox 7.0-13: the cron-job for sending the discovery data `/root/sys/proxmox_cluster.py -a adelie.domain.de -u zabbix@pve -p zabbix -t CLUSTER.domain.de -d` fails with:

`Unable to open zabbix_sender: Command '['/usr/bin/zabbix_sender', '-c/etc/zabbix/zabbix_agentd.conf', '-sCLUSTER.domain.de', '-kproxmox.nodes.discovery', '-o{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}']' returned non-zero exit status 2.`

- The -s option is set correctly and works as exptected when not using -d.
- /usr/bin/zabbix_sender is the correct path

The ouput with -v and -d:
```
{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}
Unable to open zabbix_sender: Command '['/usr/bin/zabbix_sender', '-c/etc/zabbix/zabbix_agentd.conf', '-sCLUSTER.domain.de', '-kproxmox.nodes.discovery', '-o{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}']' returned non-zero exit status 2.
```
Has there been some changes to the Proxmox-API so that the values were moved or cant be parsed correctly anymore?

**Comment by takala-jp** [comment]
Hi there,

That's interesting. I'm running it against a Proxmox 7.0-13 cluster as well and have no issues with discovery. Therefore don't believe you are running into an API issue.

Are you running the script on the Proxmox nodes or on the Zabbix server? 

Is there anything in the zabbix server logs?

**Comment by chrisgassen** [comment]
The script is running on the proxmox node.

Huh I found something interesting in the zabbix server logs: `cannot send list of active checks to "10.24.2.9": host [Hostname] not found`

Seems like our debian systems use just the hostname and not hostname.domain, we will need to change that in our ansible rules. HOWEVER changing does not seem to fix this issue. (I restarted the agent after changing the config)

**Comment by takala-jp** [comment]
I see, that means zabbix sender was also upgraded. Do you use the agent from the stock Debian repo or the zabbix repo?

That hostname not found error means the agent is sending the wrong hostname to the server. But for the monitoring script we define the hostname on the command line.

Can you test the following:

`/usr/bin/zabbix_sender -v -c /etc/zabbix/zabbix_agentd.conf -s CLUSTER.domain.de -k promox.cluster.quorate -o 1`

**Comment by chrisgassen** [comment]
We use the stock debian repo for zabbix. EDIT: the corresponding Debian version is 11.1 (bullseye) with zabbix-agent 5.0.8

`Response from "10.24.2.25:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000058"
Response from "zabbix.domain.de:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000042"
sent: 1; skipped: 0; total: 1`

**Comment by takala-jp** [comment]
That all looks good. Let's try sending the discovery data from the command line. Can you test the following:

`/usr/bin/zabbix_sender -v -c /etc/zabbix/zabbix_agentd.conf -s CLUSTER.domain.de -k proxmox.nodes.discovery -o '{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}'`

**Comment by chrisgassen** [comment]
Ok, this worked...

`root@adelie:~# /usr/bin/zabbix_sender -v -c /etc/zabbix/zabbix_agentd.conf -s CLUSTER.domain.de -k proxmox.nodes.discovery -o '{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}'
Response from "10.24.2.25:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000136"
Response from "zabbix.domain.de:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000057"
sent: 1; skipped: 0; total: 1`

I will try to look into this a little more tomorrow when I've got some more time on my hands. Currently my guess is that somehow the script does some formatting wrong.

**Comment by chrisgassen** [comment]
did some more testing and changed the block from line 146 onwards in the "proxmox_cluster.py" script to the following to get some more output:
```
        #result = subprocess.check_output([
        result = subprocess.run([
            args.zsend, "-c" , args.config, "-s" , args.target,
            "-k" , "proxmox.nodes.discovery", "-o" , str(discovery_data)
         ], capture_output=True)
```

The output then is: 

```
{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}
CompletedProcess(args=['/usr/bin/zabbix_sender', '-c', '/etc/zabbix/zabbix_agentd.conf', '-s', 'CLUSTER.domain.de', '-k', 'proxmox.nodes.discovery', '-o', '{"data": [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}'], returncode=2, stdout=b'Response from "10.24.2.25:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000051"\nResponse from "zabbix.domain:10051": "processed: 1; failed: 0; total: 1; seconds spent: 0.000045"\nsent: 1; skipped: 0; total: 1\n', stderr=b'')
```

Zabbix sender seems to get a returncode of 2 for some reason. This usually should only happen when some of the values failed. Here 1 of 1 values are processed and none failed...

Is the discovery type "zabbix trapper" correct for the discovery configuration?

**Comment by chrisgassen** [comment]
I now deleted the template from my cluster in zabbix and re-added it. The issue still persists.

Although ~~there now is an error message in the discovery tab when using the manual zabbix-sender command: `Invalid discovery rule value: cannot parse as a valid JSON object: invalid object name at: [{"{#NODE}": "chinstrap"}, {"{#NODE}": "adelie"}, {"{#NODE}": "emperor"}]}
The status says Not supported.~~ the error code is still 2, the data (items) still gets readded.

**Comment by takala-jp** [comment]
Hi Chris, TUBEOF was experiencing the same issue and suggested a fix/workaround in [PR28](https://github.com/takala-jp/zabbix-proxmox/pull/28). I'll test that before merging it at some stage.

**Comment by chrisgassen** [comment]
Thanks for notifying me of this. I recently did a similar "fix" (b4bec8ee207fdcc13b4d50b4de5397b3beda481f) but also added a small errorcode check back in. This should do it for now, the discovery works, so I dont really see a problem with it. Lets see if some day zabbix-sender gets fixed... :D
